File: Sources/LiveViewNativeCharts/Types/AnnotationPosition.swift
//
//  AnnotationPosition.swift
//
//
//  Created by Carson Katri on 5/16/24.
//

import Charts
import LiveViewNative
import LiveViewNativeStylesheet
import LiveViewNativeCore

/// See [`Charts.AnnotationPosition`](https://developer.apple.com/documentation/charts/AnnotationPosition) for more details.
///
/// Possible values:
/// * `.automatic`
/// * `.overlay`
/// * `.top`
/// * `.bottom`
/// * `.leading`
/// * `.trailing`
/// * `.topLeading`
/// * `.topTrailing`
/// * `.bottomLeading`
/// * `.bottomTrailing`
extension AnnotationPosition {
    @ASTDecodable("AnnotationPosition")
    enum Resolvable: @preconcurrency AttributeDecodable, StylesheetResolvable, @preconcurrency Decodable {
        case automatic
        case overlay
        case top
        case bottom
        case leading
        case trailing
        case topLeading
        case topTrailing
        case bottomLeading
        case bottomTrailing
    }
}

extension AnnotationPosition.Resolvable {
    func resolve<R: RootRegistry>(on element: ElementNode, in context: LiveContext<R>) -> AnnotationPosition {
        switch self {
        case .automatic:
            return .automatic
        case .overlay:
            return .overlay
        case .top:
            return .top
        case .bottom:
            return .bottom
        case .leading:
            return .leading
        case .trailing:
            return .trailing
        case .topLeading:
            return .topLeading
        case .topTrailing:
            return .topTrailing
        case .bottomLeading:
            return .bottomLeading
        case .bottomTrailing:
            return .bottomTrailing
        }
    }
    
    init(from attribute: Attribute?, on element: ElementNode) throws {
        switch attribute?.value {
        case "automatic":
            self = .automatic
        case "overlay":
            self = .overlay
        case "top":
            self = .top
        case "bottom":
            self = .bottom
        case "leading":
            self = .leading
        case "trailing":
            self = .trailing
        case "topLeading":
            self = .topLeading
        case "topTrailing":
            self = .topTrailing
        case "bottomLeading":
            self = .bottomLeading
        case "bottomTrailing":
            self = .bottomTrailing
        default:
            throw AttributeDecodingError.badValue(Self.self)
        }
    }
}

File: lib/live_view_native_swift_ui_charts/types/annotation_position.ex
defmodule LiveViewNativeSwiftUiCharts.Types.AnnotationPosition do
  use LiveViewNativePlatform.Modifier.Type
  def type, do: :string

  @positions_atom [
    :automatic,
    :bottom,
    :bottom_leading,
    :bottom_trailing,
    :leading,
    :overlay,
    :top,
    :top_leading,
    :top_trailing,
    :trailing,
  ]

  @positions_string [
    "automatic",
    "bottom",
    "bottom-leading",
    "bottom-trailing",
    "leading",
    "overlay",
    "top",
    "top-leading",
    "top-trailing",
    "trailing",
  ]

  def cast(value) when is_atom(value) and value in @positions_atom, do: {:ok, Atom.to_string(value)}
  def cast(value) when is_binary(value) and value in @positions_string, do: {:ok, value}
  def cast(_), do: :error
end
